### ISSUE TITLE
`invalid-assignment` false positive when narrowing a value to `dict[Unknown, Unknown]` and assigning a value from another `dict[Unknown, Unknown]` to it

### ISSUE THREAD

**Issue by DetachHead** [astral-sh/ty#2943]
### Summary

```py
def foo(data: object, other: dict):
    if isinstance(data, dict):
        data["foo"] = other["foo"] # Invalid subscript assignment with key of type `Literal["foo"]` and value of type `Unknown` on object of type `Top[dict[Unknown, Unknown]]` (invalid-assignment)
```

https://play.ty.dev/4263b96e-81bf-416f-a144-6ae96f98b603

### Version

0.0.20

**Comment by carljm** [comment]
This behavior is intentional. We don't narrow to `dict[Unknown, Unknown]` from `isinstance(data, dict)`, we narrow to the top materialization of `dict[Unknown, Unknown]` (that is, the union of all possible dict types), which is not a gradual type. In the typical case, if you have a union of a specific dict type with some other types, this narrowing has the desired effect. In this code, you've opted into this strict behavior (we have no idea what the key/value types of `data` are, so it's not safe to mutate it) by typing `data` initially as `object`. `object` is not a gradual type; we are not going to introduce graduality (aka unsoundness) implicitly via narrowing. If you want gradual behavior, type `data` as `Any` or `Unknown` instead (or if that's too gradual, maybe a union of `dict[Unknown, Unknown]` with some other possible types, to keep the graduality more localized).

**Comment by carljm** [comment]
Closing in order to keep the issue tracker actionable. If you think this is not a good behavior, happy to discuss and we can reopen the issue to consider alternatives.

**Comment by DetachHead** [comment]
> This behavior is intentional. We don't narrow to `dict[Unknown, Unknown]` from `isinstance(data, dict)`, we narrow to the top materialization of `dict[Unknown, Unknown]` (that is, the union of all possible dict types), which is not a gradual type.

i don't really understand this. how is a union of all possible dict types (`Top[dict[Unknown, Unknown]]`?) different to `dict[Unknown, Unknown]`?

either way i'm happy for this issue to be closed, we are trying to remove all the `Unknown`s and `Any` from our codebase anyway

**Comment by carljm** [comment]
> how is a union of all possible dict types (`Top[dict[Unknown, Unknown]]`?) different to `dict[Unknown, Unknown]`?

The second is a gradual type and the first is not. Unions are not forgiving: if you have the union of all possible dict types, you cannot mutate it since you don't know the key and value type. Gradual types are forgiving: you can put anything into `dict[Unknown, Unknown]`, because you are allowed to treat it as any dict type, instead of having to account for all possible dict types.

It's similar to the difference between `Any` and `object`. Both can represent any Python object, but `object` is a fully static type that requires that you account for all possibilities, where `Any` is a gradual type that allows you to assume any possibility.

**Comment by DetachHead** [comment]
ah so a gradual type just means `Any`/`Unknown`?

so does `Top` basically just mean "disable the unsafe behavior of `Any`/`Unknown` of the enclosed type", and is used here to work around the fact that there's no safe way to denote the upper bound of an invariant generic? so like:

- `Top[Any]` is equivalent to `object`
- `Top[Iterable[Any]]` is equivalent to `Iterable[object]`
- `Top[Callable[[Any], Any]]` is equivalent to `Callable[[Never], object]`
- and `Top[list[Any]]` can't be represented any other way

**Comment by carljm** [comment]
That's exactly right.

**Comment by ErmoshinMA1-source** [comment]
I'm confused with the `Top[dict[Unknown, Unknown]]` construct. It essentially makes the dictionary unusable – I cannot call `__getitem__`, I cannot call `.get()`, I cannot in a way narrow the type. There is no way to safely extract a value by key without using `cast`, which defeats the purpose of static checking because it just tells the type checker "trust me".

Am I missing something, or is `cast` really the only way?

```python
        from selenium.webdriver import Chrome

        value = Chrome().find_element().get_property('attributes')
        attributes: dict[str, str] = {}
        if isinstance(value, dict):
            attributes[str(value.get('name', default='default'))] = str(value.get('name', default='default'))

# Method `__getitem__` of type `bound method Top[dict[Unknown, Unknown]].__getitem__(key: Never, /) -> object` cannot be called with key of type `Literal["name"]` on object of type `Top[dict[Unknown, Unknown]]`ty[invalid-argument-type](https://ty.dev/rules#invalid-argument-type)
```

**Comment by DetachHead** [comment]
unfortunately a subtype of `dict` could override `get` in a way that breaks your code, which is what ty is protecting against here:

```py
from typing import overload, override


def fn(value: object):
    attributes: dict[str, str] = {}
    if isinstance(value, dict):
        attributes[str(value.get('name', default='default'))] = str(value.get('name', default='default'))


class Foo(dict[int, object]):
    @overload
    def get(self, key: int, default: None = None, /) -> object | None: ...
    @overload
    def get(self, key: int, default: object, /) -> object: ...
    @override
    def get[T](self, key: int, default: object = None) -> object:
        return super().get(key + 1)


fn(Foo())  # runtime crash - TypeError: can only concatenate str (not "int") to str
```

**Comment by carljm** [comment]
I actually think that typeshed should define `dict.get` as accepting `object` keys, because its safe and valid to call `.get()` even with a key that you don't necessarily know is of a type present in the dictionary. We pushed a change [just today](https://github.com/astral-sh/ruff/pull/25831) to override this part of typeshed, so the code above should no longer error on the `.get` calls. (This will also mean the `Foo` subclass above will become a Liskov violation.)
